src/backends/interflop-mca/interflop_mca.c

```diff
@@ -186,6 +189,16 @@ static double _mca_rand(void) {
   return tinymt64_generate_doubleOO(&random_state);
 }
 
+static bool _mca_skip_eval(const int sparsity) {
+  /* Returns a bool for determining whether an operation should skip */
+  /* perturbation. false -> perturb; true -> skip. */
+  if (sparsity == 1) {
+    return false;
+  }
+  /* e.g. for sparsity=10, all random values > 0.1 = true -> no MCA*/
+  return (_mca_rand() > (1.0 / sparsity));
```

Comment:
why not using an integer rand, avoid the div and cast(s) and compute the sparsity threshold onece for all as a fraction of the unit64 (or 32 64 might be overkill) and compare to this threshold.

Reply:
```suggestion
  return (rand() % sparsity);
```

Hi @titeup, would the following work? In this case, only when the mod is zero would a perturbation take place. If we do this, should I initialize the random state with `srand` next to when we initialize the other on line 590?

Reply:
the 2 random generator (mac_rand and rand()) are independent and therefore the rand seeding has no relationship to mca_rand. However we implemented an independent mca_rand as Mersen Twister in vericarlo to make sure that we are not interfering with application that might use rand(). I believe we are in the same situation.
What could be done is use our MersenTwister output as int64 and do the modulo. There are This would work.  Look in common/tinymt64.h.  I believe  tinymt64_generate_uint64 is what you are looking for.

Reply:
I'm realizing that `rand` is biased, but I think this should work with a uniform random number generator. Do you have a recommendation?

Reply:
@titeup perfect, thanks. When evaluating with the modulo, I notice that there is some bias and across sparsity values we are not consistently within a percent or two of what we expect. I tried implementing your suggestion: pre-computing the threshold as (`UINT64_MAX/sparsity`)  and continue with a `>` comparison, but also noticed more bias in this, as far as when sparsity was 4 we still perturbed 29% of operations instead of 25% over a few hundred thousand runs. Should we try harder to make this uniform, or should I just precompute the threshold to save the division each iteration and keep the logic otherwise the same?

Reply:
One solution to avoid the division is to pass the sparsity as a percentage ratio instead of an integer, like `10 -> 0.1`.

Reply:
I ended up going with @yohanchatelain 's recommendation, which also gives us a lot of flexibility in controlling the sparsity at a finer resolution.